README.md
# janus
(Janus was a two-faced Roman god who looked forward to the future and backwards to the past.  His name gives us the name of the month `January`)

[![Clojars Project](https://img.shields.io/clojars/v/com.hapgood/janus.svg)](https://clojars.org/com.hapgood/janus)

[![](https://cljdoc.org/badge/com.hapgood/janus)](https://cljdoc.org/d/com.hapgood/janus/CURRENT)

[![Build Status](https://travis-ci.org/cch1/janus.svg?branch=master)](https://travis-ci.org/cch1/janus)

A Clojure routing library.  This is my second initiative on this front -the first as a collaborator on the `wire` library.  Inspiration has been taken from:

 * [wire](https://github.com/mwmitchell/wire)
 * [bidi](https://github.com/juxt/bidi)
 * [compojure](https://github.com/weavejester/compojure)
 * [pedestal](https://github.com/pedestal/pedestal)

## Context
For something that web developers understand intuitively, the exact interpretation of "URL"s is rife with misunderstandings, competing standards and imcompatible interpretations.  In an effort to be explicit, for janus a URI is the [RFC 3986](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax)-compliant name by which a local resource is referenced on the web.

There are two primary requirements for a router:

1. Identify inbound routes -this could be satisfied by a tree of identifiers associated with regular expressions.
2. Generate outbound routes -this could be satisfied by a tree of identifiers associated with `format` strings.

These are the core functions of a "pure" routing engine and they should be powerful and generally uncompromised by secondary concerns.

## Design Goals
 1. The path components of URIs are assumed (per [RFC 3986](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax)) to be a string of `/`-separated and [url-encoded](https://en.wikipedia.org/wiki/Percent-encoding) segments (note that percent-encoding is not the same thing as RFC-compatible url-encoding, notably in the treatment of `+`).
 1. The route abstraction handles encoding and destructuring transparently; it models the URI as a sequence of URL-decoded string segments.
 1. Route identification is decomplected from dispatching to the handler.
 1. For all but the most complex routes, pure data can represent the entire route definition.
 1. Forward (generation) and backward (identification) routing are (with appropriate care) inverses of each other.
 1. Protocols are used for most conceptual tasks, e.g. segment matching/generation and dispatching.
 1. Dispatching is based on route matching only.  There is no support for dispatching based on other attributes of an HTTP request, such as method.  Other libraries, such as [liberator](https://github.com/clojure-liberator/liberator), do an excellent job of managing HTTP method-based processing _after_ routing.
 1. Routing works the same in Clojure and Clojurescript.
 1. A single compact data structure should represent all that is necessary for matching, generating and even dispatching.  For common route components (constant strings, leaf nodes, etc), the syntax should be particularly compact.
 1. Both forward and backwards routing can be performed in the scope of a parent route.

## Comparisons
 * Compared to Compojure, janus offers routes-as-data, Clojurescript compatibility, invertible routes, independent route identification and dispatching, and protocol-based extensibility.   On the downside, janus offers no support for dispatching based on HTTP method.
 * Compared to bidi, janus offers (generally) invertible routes, independent route identification and dispatching, and a higher-level abstraction that deals with encoding and destructuring.  On the downside, janus offers no support for dispatching based on HTTP method.
 * Compared to wire, janus offers invertible routes, Clojurescript compatibility and protocol-based extensibility.  On the downside, janus offers no support for dispatching based on HTTP method.

## Usage

See janus.route and janus.ring namespaces for usage notes.

### Route Format

The canonical data format for a route is a recursive tuple of two elements, the second of which is a three-tuple:

    [identifiable [as-segment dispatchable routes]]

The outer sequence can be viewed as a pair, and is satisfied by a Clojure map entry.  In fact, maps are the preferred way of representing a collection of mutually-exclusive routes:

    {identifiable1 [as-segment1 dispatchable1 routes1]
     identifiable2 [as-segment2 dispatchable2 routes2]
	 ...
     identifiableN [as-segmentN dispatchableN routesN]}

Only when the identification of route segments is not guaranteed to be unique does the order of routes become meaningful.  At that point, routes can be represented by a sequence of pairs:

    [[identifiable1 [as-segment1 dispatchable1 routes1]
     [identifiable2 [as-segment2 dispatchable2 routes2]
	 ...
     [identifiableN [as-segmentN dispatchableN routesN]]

Routes are organized as a tree: there is one root route (which can most easily be represented as a vector tuple) and all child routes have exactly one parent.

The elements of a route are:

1. `identifiable` is any instance of `clojure.lang.Named`.  This includes keywords and symbols.  Keywords should be used when the matching of the route yields useful information.  Symbols should be used for "constant" routes.  This convention is leveraged in janus' ring support namespace.
2. `as-segment` is anything satisfying `janus.route.AsSegment`, which includes strings, keywords, regexes, functions and others.  It is easy to extend AsSegment to accommodate custom interpretations.  The role of `as-segment` is twofold: to match inbound route segements (yielding route parameters) and to generate outbound route segements.  The default semantics of each are as follows:
   * `javal.lang.String`: Matches itself only, returning itself as a route parameter.  Generates itself always.
   * `clojure.lang.Keyword`: Matches its name only, returning its name as a route parameter.  Generates its name always.
   * `java.lang.Boolean`: Matches when true, never when false, and returns inbound segment.  Generates its single string parameter (such as the inbound segment) as the outbound segment.
   * `java.util.regex.Pattern`: Matches when regex matches inbound segment and returns result of regex match as route parameter(s).  Generates either the single result of matching inbound segment or the concatenation of a capture group-matched result.
   * `clojure.lang.Fn`: Matches when function applied to inbound segment returns a truthy result and returns the result as the route parameter.  Generates that same result as the outbound route segment.
   * `clojure.lang.PersistentVector`: Matches when first element (an AsSegment) matches.  Generates what second element (an AsSegment) generates.  Note that the result of matching with the first element becomes the route parameter(s) that are provided as arguments to the second.

3. `dispatchable` is anything satisfying `janus.route.Dispatchable`, which includes functions, vars and instances of `clojure.lang.Named`.  FIXME: Document better, why is protocol not enforced specifically during normalization?
4. `routes` is a recursive seqable collection of child routes.

While the canonical format for routes is useful for understanding the full capabilities of routing in janus, there are many abbreviated representations that are acceptable.  There is no run-time performance penalty for expressing routes in any abbreviated format as all routes are conformed to data types during router construction.

* Only the `identifiable` is required to represent a route.  If the second element of the pair is missing or nil, default values are assigned based on the identifiable.  For example:

        [:root nil] => [:root [nil :root {}]]
        [:root] => [:root [nil :root {}]]
   Of course suppressing the second element is not possible when using a map to represent a sequence of routes.

* If the second element of the route pair is missing components, they are either inferred from the identifier or default values are supplied.

## License

Copyright © 2020 Chris Hapgood

Distributed under the Eclipse Public License version 1.0.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).  In the spirit of Rich Hickey's "Spec-ulation" keynote address at the 2016 Clojure/conj, no backwards-incompatible changes will be made to any public var after version 1.0.0.

Note that versioning is managed using [garamond](https://github.com/workframers/garamond).

## [Unreleased](https://github.com/cch1/janus/compare/v1.4.5...HEAD)
## [1.4.5](https://github.com/cch1/janus/compare/v1.4.4...v1.4.5)
### Changed
- Adjust CljDoc badge
## [1.4.4](https://github.com/cch1/janus/compare/v1.4.3...v1.4.4)
### Updated
- Update dependencies
### Changed
- Overhaul Makefile and other tooling
## [1.4.3](https://github.com/cch1/janus/compare/v1.4.2...v1.4.3)
### Fixed
- Add badges, tweak developer docs and, once more, try to fix docs on [cljdoc](https://cljdoc.org/d/com.hapgood/janus/1.4.3).
## [1.4.2](https://github.com/cch1/janus/compare/v1.4.1...v1.4.2)
### Fixed
- Correct general docs on [cljdoc](https://cljdoc.org/d/com.hapgood/janus/1.4.2).
## [1.4.1](https://github.com/cch1/janus/compare/v1.4.0...v1.4.1)
### Fixed
- Correct API docs on [cljdoc](https://cljdoc.org/d/com.hapgood/janus/1.4.1).
## [1.4.0](https://github.com/cch1/janus/compare/v1.3.2...v1.4.0)
### Changed
- Added a groupId (com.hapgood) where previously there was none.
- Updated dependencies
- Transitioned exclusively to clojure.tools.deps.alpha tooling ecosystem (no impact on artifact though).
### Fixed
- Enumerated dependency on `ring-codec`.
## [1.3.2](https://github.com/cch1/janus/compare/v1.3.1...v1.3.2)
### Fixed
- Properly decode/encode `+` characters in the path.
## [1.3.1](https://github.com/cch1/janus/compare/v1.3.0...v1.3.1)
### Fixed
- Specify the character set in text responses to mitigate vulnerability.
## [1.3.0](https://github.com/cch1/janus/compare/v1.2.0...v1.3.0)
### Added support for the Ring async request model
## [1.2.0](https://github.com/cch1/janus/compare/v1.1.0...v1.2.0)
### Added
- Clojurescript support.
- Support distinct responses for the semantically distinct "route not found" versus "no handler available for the (found) route".
- Support for Clojure CLI during development and consumption as a Git artifact.
- Update to latest version of core.match (with much support for Clojurescript).
## [1.1.0](https://github.com/cch1/janus/compare/v1.0.0...v1.1.0)
### Added
- Implementation of pprint dispatch for better/safer printing in newer CIDER repl.
## [1.0.0](https://github.com/cch1/janus/compare/v0.7.0...v1.0.0) - 2019-02-23
### Added
- Tagged literal support for Route and RecursiveRoute
- Equality semantics (from Object.equals) for Route and RecursiveRoute
### Changed
- Refactor use of protocols and types for simplicity
### Removed
- Remove deprecated janus.ring/make-identifier
### Fixed
- Semantic failure on Boolean AsSegment implementation
## [0.7.0](https://github.com/cch1/janus/compare/v0.6.1...v0.7.0) - 2019-02-18
### Added
- Conform child routes to sequentials instead of leaving as seqable associatives.
### Changed
- Update dependencies.
### Removed
### Fixed
## [0.6.1](https://github.com/cch1/janus/compare/v0.6.0...v0.6.1) - 2017-05-11
### Added
- Route error handling with multi-method.
### Changed
- Update dependencies.
### Removed
### Fixed
## [0.6.0](https://github.com/cch1/janus/compare/v0.5.0...v0.6.0) - 2017-02-26
### Added
- Support returning generalized path.
### Changed
- Augment request with route-params early (when identifying route)
### Removed
### Fixed
- Return nil when navigating to the parent of the root.
## [0.5.0](https://github.com/cch1/janus/compare/v0.4.0...v0.5.0) - 2017-02-26
### Added
- Custom printing of Router type to suppress.
- Add support for `parent` method in Routable protocol.
- Add `Content-Type` header for 404/Not-Found response.
### Changed
- Only merge route-params with keyword identifiers into `:params`
- Adjust string and regex AsSegment behavior
### Removed
### Fixed
- Fixed bug in `root` method of Router's implementation of Routable protocol.
## [0.4.0](https://github.com/cch1/janus/compare/v0.3.1...v0.4.0) - 2017-02-22
### Added
### Changed
### Removed
### Fixed
- Assoc route params onto request at moment of dispatch, not earlier.
- Treat vars like fns during route normalization.
## [0.3.1](https://github.com/cch1/janus/compare/v0.3.0...v0.3.1) - 2017-02-22
### Added
- Unidentifiable routes are properly handled and ultimately result in a Ring 404 Not-Found response.
### Changed
- Ring identifier middleware takes an instance of janus.route.Router, not routes.
### Removed
### Fixed
## [0.3.0](https://github.com/cch1/janus/compare/v0.2.0...v0.3.0) - 2017-02-22
### Added
### Changed
- Prioritize handlers fn in normalization
### Removed
### Fixed
## [0.2.0](https://github.com/cch1/janus/compare/v0.1.0...v0.2.0) - 2017-02-21
### Added
- Allow dispatchable/handler directly in route tree.
### Changed
- Abstract routing into two primary protocols and implement with janus.route.Router type
- Use zippers for managing navigation through route tree.
## [0.1.0](https://github.com/cch1/janus/compare/...v0.1.0) - 2017-02-08
### Added
- Initial functionality for identifying & generating, plus basic ring identify/dispatching handler


test/janus/core_test.clj
(ns janus.core-test
  (:require [clojure.test :refer :all]
            [janus.core :refer :all]))

(deftest a-test
  (testing "works"
    (is true)))


test/janus/ring_test.clj
(ns janus.ring-test
  (:require [janus.ring :refer :all]
            [janus.route :as route]
            [ring.mock.request :as mock]
            [clojure.test :refer :all]))

(deftest handle-invalid-uri
  (let [router (janus.route/router [:R {:a "foo" 'a "bar"}])
        handler (-> (fn echo
                      ([request] request)
                      ([request respond raise] (respond request)))
                    (wrap-identify router))
        response (promise)
        respond #(deliver response %)
        exception (promise)
        raise #(deliver exception %)]
    (testing "Invalid URI returns response with status 400"
      (let [request (-> (mock/request :get "/")
                        (assoc :uri "/[")) ; have to force in an invalid URI ... mock will not tolerate it
            response (handler request)]
        (is (= 400 (response :status)))))
    (testing "Invalid URI returns async response with status 400"
      (let [request (-> (mock/request :get "/")
                        (assoc :uri "/["))]
        (handler request respond raise)
        (is (= 400 (@response :status)))))))

(deftest handle-identification-exceptions
  (let [router (janus.route/router [:R {:a "foo" 'a "bar"}])
        response (promise)
        respond #(deliver response %)
        exception (promise)
        raise #(deliver exception %)
        exception-handler (fn [& _] (throw (Exception. "WTF?")))
        handler (-> (fn echo
                      ([request] request)
                      ([request respond raise] (respond request)))
                    (wrap-identify router exception-handler))]
    (testing "Unusual exceptions are handled gracefully"
      (let [request (-> (mock/request :get "/")
                        (dissoc :uri))]
        (is (thrown? Exception (handler request)))
        (handler request respond raise)
        (is (not (realized? response)))
        (is (instance? Exception @exception))))))

(deftest add-route-identifier
  (let [router (janus.route/router [:R {:a "foo" 'a "bar"}])
        handler (-> (fn echo
                      ([request] request)
                      ([request respond raise] (respond request)))
                    (wrap-identify router))
        response (promise)
        respond #(deliver response %)
        exception (promise)
        raise #(deliver exception %)]
    (testing "Identified route and route-params are available in sync request"
      (let [request (mock/request :get "/foo")
            response (handler request)]
        (is (instance? janus.route.Router (response :janus.ring/router)))
        (is (= {:a "foo"} (response :route-params)))))
    (testing "Identified route and route-params are available in async request"
      (let [request (mock/request :get "/foo")]
        (handler request respond raise)
        (is (instance? janus.route.Router (@response :janus.ring/router)))
        (is (= {:a "foo"} (@response :route-params)))))
    (testing "Route parameters are only available for keyword identifiers" ; hmmmm...
      (let [request (mock/request :get "/bar")
            response (handler request)]
        (is (instance? janus.route.Router (response :janus.ring/router)))
        (is (= {'a "bar"} (response :route-params)))))
    (testing "Unidentifiable route behaves gracefully"
      (let [request (mock/request :get "/qux")
            response (handler request)]
        (is (nil? (response :janus.ring/router)))
        (is (nil? (response :route-params)))))))

(def handler (fn ([r] [(-> r :janus.ring/router route/path) (-> r :params)])
               ([r respond raise] (respond [(-> r :janus.ring/router route/path) (-> r :params)]))))

(deftest dispatch-on-route
  (let [request (assoc (mock/request :get "/foo") :params {:x "!"})
        routes [:root [nil :R {:a [#"a(?:.*)" :a]
                               :b [#"b(?:.*)" 'b]
                               :c [#"c(?:.*)" #'handler]
                               :d [#"d(?:.*)" handler]}]]
        router (route/router routes)
        dispatch-table {:a handler 'b handler}]
    (testing "dispatch on keyword"
      (let [router (route/identify router "/aX")
            request (assoc request :janus.ring/router router :route-params (into {} (route/parameters router)))
            response (promise)
            respond #(deliver response %)
            exception (promise)
            raise #(deliver exception %)]
        (is (= ["/aX" {:x "!" :a "aX"}] ((make-dispatcher dispatch-table) request)))
        ((make-dispatcher dispatch-table) request respond raise)
        (is (= ["/aX" {:x "!" :a "aX"}] @response))))
    (testing "dispatch on symbol"
      (let [router (route/identify router "/bX")
            request (assoc request :janus.ring/router router :route-params (into {} (route/parameters router)))
            response (promise)
            respond #(deliver response %)
            exception (promise)
            raise #(deliver exception %)]
        (is (= ["/bX" {:x "!" :b "bX"}] ((make-dispatcher dispatch-table) request)))
        ((make-dispatcher dispatch-table) request respond raise)
        (is (= ["/bX" {:x "!" :b "bX"}] @response))))
    (testing "dispatch on var"
      (let [router (route/identify router "/cX")
            request (assoc request :janus.ring/router router :route-params (into {} (route/parameters router)))
            response (promise)
            respond #(deliver response %)
            exception (promise)
            raise #(deliver exception %)]
        (is (= ["/cX" {:x "!" :c "cX"}] ((make-dispatcher) request)))
        ((make-dispatcher dispatch-table) request respond raise)
        (is (= ["/cX" {:x "!" :c "cX"}] @response))))
    (testing "dispatch on function"
      (let [router (route/identify router "/dX")
            request (assoc request :janus.ring/router router :route-params (into {} (route/parameters router)))
            response (promise)
            respond #(deliver response %)
            exception (promise)
            raise #(deliver exception %)]
        (is (= ["/dX" {:x "!" :d "dX"}] ((make-dispatcher) request)))
        ((make-dispatcher dispatch-table) request respond raise)
        (is (= ["/dX" {:x "!" :d "dX"}] @response))))
    (testing "Unidentified route triggers not found response"
      (let [request (assoc request :janus.ring/router nil)
            response (promise)
            respond #(deliver response %)
            exception (promise)
            raise #(deliver exception %)]
        (is (= {:status 404 :body "Not Found" :headers {"Content-Type" "text/plain; charset=US-ASCII"}}
               ((make-dispatcher) request)))
        ((make-dispatcher dispatch-table) request respond raise)
        (is (= {:status 404 :body "Not Found" :headers {"Content-Type" "text/plain; charset=US-ASCII"}}
               @response))))
    (testing "Unimplemented handler triggers not implemented response"
      (let [router (route/identify router "/aX")
            request (assoc request :janus.ring/router router :route-params {})
            response (promise)
            respond #(deliver response %)
            exception (promise)
            raise #(deliver exception %)]
        (is (= {:status 501 :body "Not Implemented" :headers {"Content-Type" "text/plain; charset=US-ASCII"}}
               ((make-dispatcher) request)))
        ((make-dispatcher) request respond raise)
        (is (= {:status 501 :body "Not Implemented" :headers {"Content-Type" "text/plain; charset=US-ASCII"}}
               @response))))))

(deftest combined-middleware
  (let [request (assoc (mock/request :get "/aX") :params {:x "!"})
        routes [:root [nil :R {:a [#"a(?:.*)" :a]
                               :b [#"b(?:.*)" 'b]
                               :c [#"c(?:.*)" #'handler]
                               :d [#"d(?:.*)" handler]}]]
        router (route/router routes)
        dispatch-table {:a handler 'b handler}
        handler (wrap-identify (make-dispatcher dispatch-table) router)
        response (promise)
        respond #(deliver response %)
        exception (promise)
        raise #(deliver exception %)]
    (is (= ["/aX" {:x "!" :a "aX"}] (handler request)))
    (handler request respond raise)
    (is (= ["/aX" {:x "!" :a "aX"}] @response))))


project.clj
(defproject janus :lein-v
  :description "A rethink of Clojure routing"
  :url "https://github.com/cch1/janus"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :plugins [[com.roomkey/lein-v "7.0.0"]
            [lein-doo "0.1.10"]
            [lein-cljsbuild "1.1.7"]]
  :middleware [lein-v.plugin/middleware]
  :release-tasks [["vcs" "assert-committed"]
                  ["v" "update"]
                  ["vcs" "push"]
                  ["deploy" "clojars"]]
  :min-lein-version "2.5.2"
  :dependencies [[org.clojure/clojure "1.10.1"]
                 [org.clojure/clojurescript "1.10.597"]
                 [org.clojure/core.match "1.0.0"]]
  :managed-dependencies [[com.google.errorprone/error_prone_annotations "2.3.4"]
                         [com.google.code.findbugs/jsr305 "3.0.2"]
                         [commons-codec "1.14"]]
  :profiles {:dev {:dependencies [[ring/ring-mock "0.4.0"]
                                  [com.bhauman/figwheel-main "0.2.3" :exclusions [org.clojure/clojurescript]]]}}
  :doo {:build "test"
        :alias {:default [:phantom]}}
  :cljsbuild {:builds [{:id "test" ; https://lambdaisland.com/episodes/testing-clojurescript
                        :source-paths ["src" "test"]
                        :compiler {:output-to "resources/public/js/testable.js"
                                   :output-dir "target"
                                   :main janus.runner
                                   :optimizations :none}}]}
  :aliases {"omnitest" ^{:doc "Run Clojure and ClojureScript tests"} ["do" "test" ["doo" "phantom" "once"]]})


src/janus/ring.clj
(ns janus.ring
  "Identify routes from Ring-compatible requests and dispatch to Ring-compatible handlers"
  (:require [janus.route :as route]))

(let [response {:status 404 :body "Not Found" :headers {"Content-Type" "text/plain; charset=US-ASCII"}}]
  (defn not-found
    ([_] response)
    ([_ respond _] (respond response))))

(let [response {:status 501 :body "Not Implemented" :headers {"Content-Type" "text/plain; charset=US-ASCII"}}]
  (defn not-implemented
    ([_] response)
    ([_ respond _] (respond response))))

(extend-protocol janus.route/Dispatchable
  nil
  (dispatch* [this args] (let [dispatch-table (first args)]
                           (route/dispatch* (get dispatch-table this not-found) args)))
  clojure.lang.Fn
  (dispatch* [this args] (apply this (rest args))) ; this is the Ring protocol for sync and async
  clojure.lang.Var
  (dispatch* [this args] (route/dispatch* (deref this) args))
  Object
  (dispatch* [this args] (let [dispatch-table (first args)]
                           (route/dispatch* (get dispatch-table this not-implemented) args))))

(defn- dispatch-request
  [{:keys [route-params params] :as request}]
  (if (and params route-params)
    (update request :params merge (into {} (filter (fn [[k _]] (keyword? k)))
                                        route-params))
    request))

(defn make-dispatcher
  ([] (make-dispatcher {}))
  ([dispatch-table]
   (fn dispatch
     ([{router ::router :as request}]
      (let [request (dispatch-request request)]
        (route/dispatch* router [dispatch-table request])))
     ([{router ::router :as request} respond raise]
      (let [request (dispatch-request request)]
        (route/dispatch* router [dispatch-table request respond raise]))))))

(defmulti exception-handler "Handle exceptions that don't allow routing to execute" class)

(defmethod exception-handler java.net.URISyntaxException
  [_] {:status 400 :body "Invalid URI Syntax" :headers {"Content-Type" "text/plain; charset=US-ASCII"}})

(defmethod exception-handler :default
  [e] (throw e))

(defn- identify-request
  [router {:keys [uri path-info] :as request}]
  (let [r (route/identify router (or path-info uri))
        route-params (when r (into {} (route/parameters r)))
        request (if route-params (assoc request :route-params route-params) request)]
    (assoc request ::router r)))

(defn wrap-identify
  "Create Ring middleware to identify the route of a request based on `:path-info` or `:uri`"
  ([handler router] (wrap-identify handler router exception-handler))
  ([handler router exception-handler]
   {:pre [(instance? janus.route.Router router)]}
   (fn identifier
     ([request]
      (try (handler (identify-request router request))
           (catch Exception e (exception-handler e))))
     ([request respond raise]
      (try (try (handler (identify-request router request) respond raise)
                (catch Exception e (respond (exception-handler e))))
           (catch Throwable e (raise e)))))))
